src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -164,81 +188,122 @@ void disableRateLimiting() {
+                        subscribe.getTopic(),
+                        QualityOfService.getEnumValueFromInteger(subscribe.getQos().getValue()))
                 .thenApply((i) -> new SubscribeResponse(null, subscribe.getQos().getValue(), null));
     }
 
     @Override
     public CompletableFuture<PubAck> publish(Publish publish) {
-        return publish(new MqttMessage(publish.getTopic(), publish.getPayload(),
-                        QualityOfService.getEnumValueFromInteger(publish.getQos().getValue()), publish.isRetain()),
-                QualityOfService.getEnumValueFromInteger(publish.getQos().getValue()), publish.isRetain()).thenApply(
-                (i) -> new PubAck(PubAckPacket.PubAckReasonCode.SUCCESS.getValue(), null, null));
+        return publish(
+                        new MqttMessage(
+                                publish.getTopic(),
+                                publish.getPayload(),
+                                QualityOfService.getEnumValueFromInteger(
+                                        publish.getQos().getValue()),
+                                publish.isRetain()),
+                        QualityOfService.getEnumValueFromInteger(publish.getQos().getValue()),
+                        publish.isRetain())
+                .thenApply(
+                        (i) ->
+                                new PubAck(
+                                        PubAckPacket.PubAckReasonCode.SUCCESS.getValue(),
+                                        null,
+                                        null));
     }
 
-    private CompletableFuture<Integer> publish(MqttMessage message, QualityOfService qos, boolean retain) {
-        return connect().thenCompose((b) -> {
-            // Take the tokens from the limiters' token buckets.
-            // This is guaranteed to not block because we've already slept the required time
-            // in the spooler thread before calling this method.
-            transactionLimiter.acquire();
-            bandwidthLimiter.acquire(message.getPayload().length);
-            try (LockScope ls = LockScope.lock(lock)) {
-                throwIfNoConnection();
-                logger.atTrace().kv(TOPIC_KEY, message.getTopic()).kv(QOS_KEY, qos.name()).kv("retain", retain)
-                        .log("Publishing message");
-                return connection.publish(message, qos, retain);
-            }
-        });
+    private CompletableFuture<Integer> publish(
+            MqttMessage message, QualityOfService qos, boolean retain) {
+        return connect()
+                .thenCompose(
+                        (b) -> {
+                            // Take the tokens from the limiters' token buckets.
+                            // This is guaranteed to not block because we've already slept the
+                            // required time
+                            // in the spooler thread before calling this method.
+                            transactionLimiter.acquire();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Lock acquisition without timeout detected. Using locks without timeouts can lead to  indefinite blocking and service outages if the lock cannot be acquired. Use timeout-based methods like tryLock(timeout), tryAcquire(timeout), or await(timeout) instead. Learn more on : https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Lock.html#tryLock-long-java.util.concurrent.TimeUnit